Extend docs about project initialization (#1785)
software-mansion/protostar: website/docs/tutorials/08-cairo-1-support/01-understanding-cairo-packages.md

# Understanding Cairo Packages

> [//]: # (TODO intro)

Reviewer: You have a TODO here :)

## `.cairo` files

Files with `.cairo` extension contain cairo code, including Starknet contracts. A file can define multiple methods,
structures etc.

## Modules

Module consists of one or more cairo files, usually organized in a single directory. To define a module, create `.cairo`
file named like the module and define components of this module with `mod` keyword.

```
my_module/
├── a.cairo
├── b.cairo
└── c.cairo
my_module.cairo
```

> ```cairo title="my_mod.cairo"

Reviewer: ```suggestion
```cairo title="my_module.cairo"
```

mod a;
mod b;
mod c;
```

> Alternatively, modules can be defined with a file itself using

Reviewer: ```suggestion
Alternatively, modules can be defined within a file using
```

> ```cairo title="my_mod.cairo"

Reviewer: ```suggestion
```cairo title="my_module.cairo"
```

# Understanding Cairo Packages

There are several requirements that Cairo packages have to follow. Further documentation pages assume some knowledge

> of Cairo package structure. Please refer to this page for details.

Author: Reworded

## `.cairo` files

Files with `.cairo` extension contain cairo code, including Starknet contracts. A file can define multiple methods,
structures etc.

## Modules

Module consists of one or more cairo files, usually organized in a single directory. To define a module, create `.cairo`
file named like the module and define components of this module with `mod` keyword.

```
my_module/
├── a.cairo
├── b.cairo
└── c.cairo
my_module.cairo
```

```cairo title="my_module.cairo"
mod a;
mod b;
mod c;
```

Alternatively, modules can be defined within a file using

```cairo title="my_module.cairo"
mod my_module {
    fn function_a() -> ();
    // ...
}
```

## Packages

> Package is a collection of modules that also defines `cairo_project.toml` and `lib.cairo` files.

Author: I meant that, just worded it purely. Will change that.

:::info
Some other tools use name "crates" for packages.

> :::

Author: The problem is, currently official cairo 1 docs use term `crate` quite extensively while protostar and scarb use `package`. I think someone should yield eventually and we should agree on one name, but for the time being, we should at least acknowledge this difference.

> :::

Reviewer 2: I understand, I just think it would be far more understandable if you wrote something like "In Cairo 1 docs they usually use the name 'crate' which is basically the same as a package here", why don't we do this?
I mean, "some tools"? How a user may know what you had in mind? What tools? Linux commands? Node.js packages? C++ libs? Cairo, not Cairo, solidity?

> :::

Author: I can extend the info so it references cairo 1 docs explicitly. However, since `crate` is "official" term, other tools that are not made by us can also use term crate - that's why I didn't specify any tool initially.

### `cairo_project.toml`

> It is needed for the definition of `[crate_roots]`, which is a directory containing `lib.cairo`.

Author: I'm not really sure if calling it a definition of package is a good idea. 

Perhaps something like "It is required for defining a package. It specifies the name of the package. " is better description?

> It is needed for the definition of `[crate_roots]`, which is a directory containing `lib.cairo`.

Reviewer 3: > It specifies the name of the package.
sounds good

The default `cairo_project.toml` file contains only the definition of the `hello_starknet` package which is contained in
the `src` directory.

It is good practice for the package name defined here to match the top-level directory name.

```toml title="cairo_project.toml"
[crate_roots]
hello_starknet = "src"
```

### `lib.cairo`

It is the root of the package tree. Here you can define functions, declare used modules, etc.

The default one has `contracts` and `business_logic` module declarations:

```cairo title="lib.cairo"
mod business_logic;
mod contracts;

> ```

Reviewer 3: Can we provide an abstract example here instead?

### Creating and using a new modules

Suppose we wanted to create a module called `mod1` inside the `hello_starknet` package and use it in tests.
We want this module to only have one file `functions.cairo` containing one function defined like:

```cairo title="functions.cairo"
fn returns_three() -> felt252 {
    3
}
```

Here is our initial file structure

```
my_project/
└── hello_starknet/
    ├── src/
    │   ├── business_logic/
    │   │   └── utils.cairo
    │   ├── contracts/
    │   │   └── hello_starknet.cairo
    │   ├── business_logic.cairo
    │   ├── contracts.cairo
    │   └── lib.cairo

>     └── cairo_project.toml

Reviewer 3: I would add two modules here though, to emphasise that packages is just modules packed together

>     └── cairo_project.toml

Reviewer 2: I'm not following, sorry. I think this can be simpler 🤔 I made a suggestion that would work and is far more simple than what you proposed. I understand that what you put here is probably consistent with a template you were doing in another pr but I still hold my opinion from the original comment - as a matter of fact I don't feel it got addressed in any of your comments unfortunately :( Could you go through my comment above and address both the suggestion and what I pointed out? Thanks :)

>     └── cairo_project.toml

Reviewer 2: I see. I still think this should be done as I said before.

> I think the structure you are suggesting is too simple for the purpose of this example

That's interesting, I think the very first example here should be as simple as possible. Here's the problem: I think the simplest example ever will be easier to understand for the users and you think that enforcing good practices from the very beginning of the docs is a good thing even if we start off with a little more complicated example.

I think we could meet halfway here. What about this: We could start with the simplest possible example and comment on it like "This is the minimum you need for this to work" and then we could say "But here's a better example with good practices and basically you should go with this."? I think that would come with the benefit of quality to this doc.

>     └── cairo_project.toml

Author: On the second thought, I see your point with using minimal example here. We have an example better reflecting project structure on the Project initialization page.

```

#### Adding a new module

Here are the steps we need to take:

1. Create a `mod1` subdirectory inside `src`.
2. Create file `functions.cairo` inside `mod1` subdirectory and define your code there
3. Create `mod1.cairo` file **in the `src` directory**, with the contents of

```cairo title="mod.cairo"
mod functions;
```

4. Update the `lib.cairo` file to include `mod1`. It's contents should now look like this

```cairo title="lib.cairo"
// previous code stays
// ...
mod mod1;
```

If you followed the steps correctly, your new project structure should look like this

```
my_project/
└── hello_starknet/
    ├── src/
    │   ├── business_logic/
    │   │   └── utils.cairo
    │   ├── contracts/
    │   │   └── hello_starknet.cairo
    │   ├── mod1/  <------------------- new directory
    │   │   └── functions.cairo  <----- new file
    │   ├── business_logic.cairo
    │   ├── contracts.cairo
    │   ├── lib.cairo  <--------------- contents updated

>     │   └── mod1.cairo  <-------------- new file

Reviewer 2: How are the contents updated? What are the new file contents? This was never explained here and may be confusing for the users.

Reviewer 2: Ok, right, sorry I missed that!

    └── cairo_project.toml
```

#### Using added module

> You now use your function in the HelloStarknet contract use `hello_starknet::mod1::functions::returns_three()`.

Reviewer 3: Sections `#### Adding a new module` `#### Using added module` are unnecessarily wordy, Just say

> Single package can contain multiple modules
And give an example of the dir tree

Author: So you don't want to include the example? I think it's quite useful because it explains how exactly you declare the module in code and import it in the upper level.

Reviewer 3: I mostly don't like this part
```
1. Create a `mod1` subdirectory inside `src`.
2. Create file `functions.cairo` inside `mod1` subdirectory and define your code there
3. Create `mod1.cairo` file **in the `src` directory**, with the contents of
```
I would just provide a tree, and add info how to call between modules

Author: @Arcticae what do you think?

Reviewer: I think maks has a point, i would make sure user knows what is in which file

Author: I think removing the code examples from these files will just make this guide less useful. Imo, it's not obvious that `mod1.cairo` should (usually) just contain `mod functions;` and nothing more.

## Packages and modules names considerations

> The names must use only ASCII alphanumeric characters or `_`, and cannot be empty. It cannot also start with underscore.

Reviewer 2: ```suggestion
The names must use only ASCII alphanumeric characters or `_`, and cannot be empty. It cannot also start with an underscore.
```

> The names must use only ASCII alphanumeric characters or `_`, and cannot be empty. It cannot also start with underscore.

Reviewer 2: I think we're missing a link to a working example with the code mentioned above. There should be a possibility for users to easily see the code in a repo, they could also quickly clone it, check that it works, and play with it.

Author: This code is basically the initial template with an added module, and the template itself is not directly usable unless created with `protostar init-cairo1`. I'm not sure if this would be useful to additionally include it.

Reviewer 2: I think it always helps to be able to see the code in the GitHub tree but I leave the decision to you.

> The names must use only ASCII alphanumeric characters or `_`, and cannot be empty. It cannot also start with underscore.

Reviewer 3: ```suggestion
The names must use only ASCII alphanumeric characters or `_`, and cannot be empty. Additionally, It cannot start with underscore.
```
